### DISCUSSION TITLE
Ведение реестра платежей в строительной организации

### DISCUSSION THREAD

**Discussion by Vadimgrig2012** [totumonline/totum-ru-issues-and-discussions#244] [Как это сделать?] [answered]
Добрый день. Подскажите пожалуйста, в каких случаях какие таблицы необходимо использовать? В архитектуре все просто: 

1) Простой пользователь - у него есть доступ к форме в которой он может добавить счет (указать номер счета, подгрузить файл, указать контрагента, указать сумму и т.д.). На сколько я понимаю - это должна быть временная таблица и форма должна вытягивать поля для заполнения из временной таблицы, но я не уверен. Так же данный пользователь по нажатию на кнопку должен получать ограниченный отчет из общей таблицы-хранилища, в котором будет отображаться информация о счетах конкретно данного пользователя и статусы согласования в виде чек-бокса (согласован счет руководством или нет). Тут я уверен на 100% что это должна быть временная таблица.

Далее при сохранении данные с формы должны улетать в общую таблицу-хранилище, где будут просто храниться вся информация (кто, когда, что добавил, даты, загруженные файлы и т.д.) и + должны быть проставлены дополнительно несколько полей с чекбоксами (как раз поля согласования). Тут я на 100% уверен, что это должна быть обыкновенная таблица и с хранящихся в ней данных можно строить все остальные отчеты.

2) Пользователь-руководитель имеет право на просмотр информации о всех счетах от всех пользователей И ТАК ЖЕ ОН МОЖЕТ ОБОЗНАЧАТЬ СРОЧНОСТЬ ОПЛАТЫ СЧЕТА НАЖИМАЯ НА ЧЕКБОКС - вот тут у меня и началась запара. Не пойму - как правильно это оформить... То ли во временную таблицу вытащить данные из общей таблицы хранилища (по любому нужен пересчет в футере, т.к. руководитель должен понимать, на какую сумму у него срочные платежи), то ли как то иначе... Не понятно, какую таблицу использовать лучше....

3) Пользователь-директор - к нему в отчет должны попадать только те счета, КОТОРЫМ БЫЛА ПРИСВОЕНА СРОЧНОСТЬ ПОЛЬЗОВАТЕЛЕМ-РУКОВОДИТЕЛЕМ и эти счета директор должен ОБОЗНАЧАТЬ ЧЕКБОКСОМ СИМВОЛИЗИРУЯ О О ТОМ, ЧТО ИХ НАДО ОПЛАТИТЬ... Вот тут тоже самое - не пойму, как связать то... Через какую таблицу... В голове не вяжется логика....

4) И "последний персонаж в нашей пьесе" пользователь-бухгалтер - к нему в отчет должны попадать только те счета, КОТОРЫЕ ОТМЕТИЛ ЧЕКБОКСОМ "В ОПЛАТУ" ДИРЕКТОР... После оплаты бухгалтер должна поставить чекбоксы на те позиции, которые были оплачены, и оплаченные счета не должны в итоге ни у кого больше фигурировать... Только в общей таблице-хранилище... Тут тоже самое что и в п.2 и 3 - - - не понимаю, какую таблицу лучше использовать...

**Comment by k1erk** [comment]
Добрый. сделайте 1 расчетную в дереве. Сделайте поле в хэдере с кодом на текущего пользователя. Второе поле в хэдере тип данные и соберите selectRowList с условиями под каждую роль пользователя + условия чекбоксов true/false (можно собирать номер цикла, id поля и тд, для дальнейшего кода действия при изменении менять данные в связанных таблицах).
В строчной части поле insert данные, далее разбираете массив + нужные кнопки и действия.

**Reply by erpic-ru** [reply]
Временная таблица для формы, которая будет инсертить данные в простую таблицу - хранилище с префильтрами  расчетом суммы по префильтрам через форматирование.

Либо временная таблица с футером для директора, выгребающая данные из основной таблицы по условиям через автозаполнение.

Для бухгалтера будет вполне достаточно префильтра по роли.

Расчетная в дереве - не очень хороший вариант.

**Reply by k1erk** [reply]
> Временная таблица для формы, которая будет инсертить данные в простую таблицу - хранилище с префильтрами расчетом суммы по префильтрам через форматирование.
> 
> Либо временная таблица с футером для директора, выгребающая данные из основной таблицы по условиям через автозаполнение.
> 
> Для бухгалтера будет вполне достаточно префильтра по роли.
> 
> Расчетная в дереве - не очень хороший вариант.

Почему расчетная в дереве не очень хороший вариант, когда она будет строиться каждый раз с меньшим кол-вом данных от selectRowList &
Мало ли я что то не понимаю и не учитываю, и как будет выглядеть экспорт csv для начальства при способе через форматирование ?

**Reply by erpic-ru** [reply]
Для одного пользователя особой разницы нет, но 2+ пользователя ушатают эту таблицу в первый же момент.
Про CSV вроде задачи не было, но если есть, то нужно делать отдельный интерфейс на временной таблице

**Reply by k1erk** [reply]
> Для одного пользователя особой разницы нет, но 2+ пользователя ушатают эту таблицу в первый же момент. Про CSV вроде задачи не было, но если есть, то нужно делать отдельный интерфейс на временной таблице

Ну насчет ушатаем, я сомневаюсь т.к зависит от факторов железа, аппетита, задач и возможностей.

**Reply by ttm-alexeypolunin** [reply]
Я вот считаю, что эта задача для одной простой таблицы и префильтров. Потом к ней можно подцепить временные на добавление для большего удобства.

Расчетная в дереве вообще очень опасная штука — надо стараться всеми средствами избегать такого типа.

Описана хорошая законченная задача, я через пару дней по ней сделаю небольшой мастеркласс для начинающих.

**Reply by ttm-alexeypolunin** [reply]
Вот собственно в чем проблема расчетной в дереве: https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/33

**Reply by Vadimgrig2012** [reply]
Спасибо, Алексей. Сейчас именно из одной простой таблицы и пытаюсь это сотворить.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### Мастеркласс: https://ru.docs.totum.online/workshop-1-1

### Видео на YouTube: https://youtu.be/d2DTv44RUAI

#### Файл примера: https://raw.githubusercontent.com/totumonline/ttm-samples/main/samples/ttm_masterclass_1.json.gz.ttm

Инструкции по установке примеров: https://github.com/totumonline/ttm-samples
